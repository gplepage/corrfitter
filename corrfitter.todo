- Add a median() function (like in lsqfit.Accumulator) to Dataset.

- Fix assemble(): doesn't work now if template contains reg-expression symbols (like +) since the template elements are used as regular expressions in one place (grep()) but later as literal strings. Must be all the same.