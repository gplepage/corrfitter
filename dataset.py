#!/usr/bin/env python
# encoding: utf-8
"""
dataset.py --- simplified replacement for old module; for legacy purposes only 
               (use gvar.dataset for new stuff).
"""
# Created by G. Peter Lepage, Cornell University, on 2012-05-22.
# Copyright (c) 2010-2012 G. Peter Lepage.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version (see <http://www.gnu.org/licenses/>).
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

import gvar
import numpy

def flexargfcn(f):
    """ decorator for Dataset analysis functions """
    def newf(self,arg=None):
        if arg is not None:
            if (hasattr(arg,"__hash__") and arg.__hash__ is not None 
                and arg in self): 
                return f(self,self[arg])
            else:
                ans = {}
                for k in arg:
                    ans[k] = f(self,self[k]) 
                return ans
        else:
            ans = {}
            for k in self:
                ans[k] = f(self,self[k]) 
            return ans
    ##
    newf.__doc__ = f.__doc__
    return newf
##

class Dataset(gvar.dataset.Dataset):
    """ legacy class --- don't use otherwise """
    def __init__(self, *args,**kargs):
        if 'bstrap' in kargs:
            bstrap = kargs['bstrap']
            del kargs['bstrap']
        else:
            bstrap = False
        super(Dataset, self).__init__(*args,**kargs)
        self.bstrap = bstrap
    ##
    def copy(self,dlist):
        if type(dlist) not in [list,tuple]:
            dlist = [dlist]
        for d in dlist:
            self.extend(d)
    ##
    def nmeas(self):
        dlen = [len(self[k]) for k in self]
        return min(dlen),max(dlen)
    ##
    @flexargfcn
    def avg(self,x):
        return numpy.mean(x,axis=0)
    ##
    @flexargfcn
    def median(self,x):
        return gvar.dataset.avg_data(x,median=True,spread=self.bstrap)
    ##
    @flexargfcn
    def sdev(self,x):
        if self.bstrap:
            return numpy.std(xx,axis=0)
        else:
            return numpy.std(xx,axis=0)/math.sqrt(len(xx))
    ##
    @flexargfcn
    def cov(self,d):
        oldshape = d[0].shape
        d = [di.flatten() for di in d]
        # d = d.reshape((d.shape[0],-1)) # flatten all other dimensions
        dcov = numpy.cov(d,rowvar=False,bias=True)
        if self.bstrap:
            return dcov.reshape(oldshape+oldshape)
        else:
            return dcov.reshape(oldshape+oldshape)/float(len(d))
    ##     
    def gdev(self,arg=None):
        if arg is None:
            arg = self.keys()
            strip_ans = False
        elif (hasattr(arg,'__hash__') and arg.__hash__ is not None 
              and arg in self):
            arg = [arg]
            strip_ans = True
        if len(arg)==0:
            return dict()
        dd = dict()
        for k in arg:
            dd[k] = self[k]
        ans = gvar.dataset.avg_data(dd,spread=self.bstrap)
        return ans[arg[0]] if strip_ans else ans
    ##
    def bin(self,nbin=2):
        return gvar.bin_data(self,binsize=nbin)
    ##
    def bootstrap_iter(self,n=None):
        return gvar.dataset.bootstrap_iter(self,n)
    ##
    def assemble(self,template,newtag=''):
        ans = Dataset()
        ans[newtag] = self.arrayzip(template)
        return ans
##  

def tabulate_avg(avgout,format=(6,3)):
    """ Tabulates averages and standard deviations.
        
    tabulate_avg(...) creates a nicely formatted table displaying the
    output from functions like ``dataset.Dataset.gdev``. Here ``avgout`` is
    the output. Parameter ``format`` specifies the output format:
    ``format=(N,D)`` implies that format ``'%N.Df(%Dd)'`` is used to print
    ``avg,int(10**D * std_dev)``. The table is returned as a single string,
    for printing.
    """
    table = []
    output = avgout.items()
    output.sort()
    for tag,avsd in output:
        try:
            av = avsd.mean
            sd = avsd.sdev
        except AttributeError:
            av = gvar.mean(avsd)
            sd = gvar.sdev(avsd)
        lines = ''
        line = '%15s' % str(tag)
        try:
            sdfac = 10**format[1]
            fmt = (' %'+str(format[0])+'.'+str(format[1])+
                  'f(%'+str(format[1])+'d)')
            def avgfmt(av,sd,fmt=fmt,sdfac=sdfac):
                try:
                    return fmt % (av,int(sdfac*sd+0.5))
                except:
                    return (' %g (%.4g)' % (av,sd))
            ##
        except:
            def avgfmt(av,sd):
                return (' %g (%.4g)' % (av,sd))
            ##
        na = len(av)
        if len(sd)<na:
            na = len(sd)
        if na>=1:
            for i in xrange(na):
                if len(sd.shape)==2:
                    sdi = math.sqrt(sd[i][i])
                else:
                    sdi = sd[i]
                nextfield = avgfmt(av[i],sdi)
                if (len(nextfield)+len(line))>78:
                    lines = lines + line + '\n'
                    line = ''.ljust(15) + nextfield
                else:
                    line = line + nextfield
            table.append(lines + line +'\n')
    return '\n'.join(table)
##    
    

if __name__ == '__main__':
    import gvar
    gvar.ranseed((1950,1))
    r1 = gvar.gvar(8.,1.)
    r2 = gvar.gvar([-10.,-9.],[2.,3.])
    r3 = gvar.gvar([[0.,1.],[2.,3.]],[[1.,2.],[3.,4.]])
    r3_iter = gvar.raniter(r3)
    r2_iter = gvar.raniter(r2)
    N = 1001
    d = Dataset(bstrap=False)
    for x in range(N):
        d.append('x',r3_iter.next())
    for x in range(N):
        d.append('y',r2_iter.next())
    d2 = Dataset(bstrap=False)
    for x in range(N):
        d2.append('z',r1())
    d.copy(d2)
    med = d.gdev()
    for k in med:
        print( k,med[k])
    avg = d.avg()
    for k in avg:
        print( k,avg)
    print( d.nmeas())
    nd = d.assemble(['y','y'],'yy')
    print( nd.avg())
    nd = d.grep('x|y')
    print( nd.keys())
        
    output = """    
    y [-10.0069 +- 0.00632473 -8.98159 +- 0.00949937]
    x [[-0.00311314 +- 0.00316463 1.00027 +- 0.00634238]
     [2.00083 +- 0.00950991 2.99893 +- 0.0126476]]
    z 8.00641 +- 0.00315755
    """     # 15.5 sec with N = 100001