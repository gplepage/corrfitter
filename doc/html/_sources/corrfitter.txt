:mod:`corrfitter` - Least-Squares Fit to Correlators
=====================================================

.. moduleauthor:: G.P. Lepage <g.p.lepage@cornell.edu>

.. |CorrFitter| replace:: :class:`corrfitter.CorrFitter`
.. |Corr2| replace:: :class:`corrfitter.Corr2`
.. |Corr3| replace:: :class:`corrfitter.Corr3`
.. |BaseModel| replace:: :class:`corrfitter.BaseModel`
.. |Dataset| replace:: :class:`gvar.dataset.Dataset`
.. |GVar| replace:: :class:`gvar.GVar`

.. automodule:: corrfitter
   :synopsis: Least-squares fitter for correlators.


Correlator Model Objects
-------------------------
Correlator objects describe theoretical models that are fit to
correlator data by varying the models' parameters. 

A model object's parameters are specified through priors for the fit. A
model assigns labels to each of its parameters (or arrays of related
parameters), and these labels are used to identify the corresponding
parameters in the prior. Parameters can be shared by more than one model
object.

A model object also specifies the data that it is to model. The data is 
identified by the data tag that labels it in the input file or |Dataset|.

.. autoclass:: corrfitter.Corr2
   :members:
   
.. autoclass:: corrfitter.Corr3
   :members:
   
.. autoclass:: corrfitter.BaseModel
   :members:


|CorrFitter| Objects
---------------------
|CorrFitter| objects are wrappers for :func:`lsqfit.nonlinear_fit()` which
is used to fit a collection of models to a collection of Monte Carlo data.

.. autoclass:: corrfitter.CorrFitter
   :members:

