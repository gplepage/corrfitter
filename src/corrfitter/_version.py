__version__ = '8.2'