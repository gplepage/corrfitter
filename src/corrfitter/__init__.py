"""
This module contains tools that facilitate least-squares fits, as functions
of time ``t``, of simulation (or other statistical) data for 2-point and
3-point correlators of the form::

    Gab(t)    =  <b(t) a(0)>
    Gavb(t,T) =  <b(T) V(t) a(0)>

where ``T > t > 0``. Each correlator is modeled using |Corr2| for 2-point
correlators, or |Corr3| for 3-point correlators in terms of amplitudes for
each source ``a``, sink ``b``, and vertex ``V``, and the energies
associated with each intermediate state. The amplitudes and energies are
adjusted in the least-squares fit to reproduce the data; they are defined
in a shared prior (typically a dictionary).

An object of type |CorrFitter| describes a collection of correlators and is
used to fit multiple models to data simultaneously. Fitting multiple
correlators simultaneously is important if there are statistical
correlations between the correlators. Any number of correlators may be
described and fit by a single |CorrFitter| object.

Typical code has the structure ::

    from corrfitter import CorrFitter

    data = make_data('mcfile')          # user-supplied routine
    models = make_models()              # user-supplied routine
    N = 4                               # number of terms in fit functions
    prior = make_prior(N)               # user-supplied routine
    fitter = CorrFitter(models=models)
    fit = fitter.lsqfit(data=data, prior=prior)  # do the fit
    print_results(fit, prior, data)     # user-supplied routine

where ``make_data`` assembles the correlator data to be fit,
``make_prior`` defines Bayesian priors for the fit parameters,
``fitter.lsqfit(...)`` does the fit, and ``print_results``
writes out the rsults. Sample code for each of these routines is
given in the tutorial documentation for |CorrFitter|.
"""

# Created by G. Peter Lepage, Cornell University, on 2010-11-26.
# Copyright (c) 2010-2021 G. Peter Lepage.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version (see <http://www.gnu.org/licenses/>).
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

from ._corrfitter import * 
from ._version import __version__