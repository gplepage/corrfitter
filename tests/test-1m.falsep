(dp0
S'log(Ds:a)'
p1
cnumpy.core.multiarray
_reconstruct
p2
(cnumpy
ndarray
p3
(I0
tp4
S'b'
p5
tp6
Rp7
(I1
(I1
tp8
cnumpy
dtype
p9
(S'f8'
p10
I0
I1
tp11
Rp12
(I3
S'<'
p13
NNNI-1
I-1
I0
tp14
bI00
S'JI\xdb\xbc\n\x9b\xf8\xbf'
p15
tp16
bsS'log(Ds:ao)'
p17
g2
(g3
(I0
tp18
g5
tp19
Rp20
(I1
(I1
tp21
g12
I00
S'\x88\xecm\x16\xf4\xaa\x04\xc0'
p22
tp23
bsS'Vno'
p24
g2
(g3
(I0
tp25
g5
tp26
Rp27
(I1
(I1
I1
tp28
g12
I00
S'0\x06\xf8\x05\xf53\xe7\xbf'
p29
tp30
bsS'Vnn'
p31
g2
(g3
(I0
tp32
g5
tp33
Rp34
(I1
(I1
I1
tp35
g12
I00
S'\xc45Y\xfc\x88\x80\xe8?'
p36
tp37
bsS'log(Ds:dEo)'
p38
g2
(g3
(I0
tp39
g5
tp40
Rp41
(I1
(I1
tp42
g12
I00
S'\xd7\xf6-\xa1\xe9G\xd8?'
p43
tp44
bsS'log(etas:dE)'
p45
g2
(g3
(I0
tp46
g5
tp47
Rp48
(I1
(I1
tp49
g12
I00
S'\x07\xea\x8c\xf6\xd4\x0c\xec\xbf'
p50
tp51
bsS'log(Ds:dE)'
p52
g2
(g3
(I0
tp53
g5
tp54
Rp55
(I1
(I1
tp56
g12
I00
S'\xec\xbc\xaf\x18v\x86\xc7?'
p57
tp58
bsS'log(etas:a)'
p59
g2
(g3
(I0
tp60
g5
tp61
Rp62
(I1
(I1
tp63
g12
I00
S'\xc4\xa5A\xa7\xcdW\xf8\xbf'
p64
tp65
bs.