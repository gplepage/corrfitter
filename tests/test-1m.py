
#!/usr/bin/env python
# encoding: utf-8
"""
test-1m.py -- tests marginalization

Created by Peter Lepage on 2010-11-26.
Copyright (c) 2010/2011 Cornell University. All rights reserved.
"""

import os
from corrfitter import Corr2,Corr3,CorrFitter
from gvar import gvar,log,exp,BufferDict
from gvar.dataset import Dataset,avg_data
from numpy import array,arange

import lsqfit

lsqfit.nonlinear_fit.fmt_parameter = '%7.3f +- %7.3f'

DISPLAYPLOTS = False         # display plots at end of fitting

P0 = {}
P0[True] = dict([('log(etas:a)', array([-1.52132077])), ('log(etas:dE)', array([-0.87652893])), ('log(Ds:a)', array([-1.5379696])), ('log(Ds:dE)', array([ 0.18376377])), ('log(Ds:ao)', array([-2.59384428])), ('log(Ds:dEo)', array([ 0.37794027])), ('Vnn', array([[ 0.76905768]])), ('Vno', array([[-0.7629936]]))])
P0[False] = dict([('log(etas:a)', array([-1.52128644])), ('log(etas:dE)', array([-0.8765244])), ('log(Ds:a)', array([-1.53828898])), ('log(Ds:dE)', array([ 0.18373669])), ('log(Ds:ao)', array([-2.63514295])), ('log(Ds:dEo)', array([ 0.3728692])), ('Vnn', array([[ 0.76630814]])), ('Vno', array([[-0.71139104]]))])

try: 
    import matplotlib
except ImportError:
    DISPLAYPLOTS = False

def main():
    dfile = "coarse_3pt_etas_Ds.bint8"  # data file
    data = avg_data(Dataset(dfile))
    prior = build_prior(6)              # 6 terms in prior
    nexp = 1                            # nexp terms in fit
    for ratio in [True,False][:]:
    # ratio = True
    # for nexp in [1,2,3,4,5,6][:1]:
        fitter = CorrFitter(models=build_models(),nterm=(nexp,nexp),ratio=ratio)
        print '========================== nexp =',nexp,'  ratio =',ratio
        fit = fitter.lsqfit(data=data,prior=prior,p0=P0[ratio])
        print_results(fit,prior,data)
        print '\n\n'
    if DISPLAYPLOTS:
        fitter.display_plots()
##
      
def print_results(fit,prior,data):
    """ print out additional results from the fit """
    dEetas = exp(fit.p['log(etas:dE)'])
    print 'etas:dE =',fmtlist(dEetas[:3])
    print ' etas:E =',fmtlist([sum(dEetas[:i+1]) for i in range(3)])
    print ' etas:a =',fmtlist(exp(fit.p['log(etas:a)']))
    dEDs = exp(fit.p['log(Ds:dE)'])
    print
    print 'Ds:dE =',fmtlist(dEDs[:3])
    print ' Ds:E =',fmtlist([sum(dEDs[:i+1]) for i in range(3)])
    print ' Ds:a =',fmtlist(exp(fit.p['log(Ds:a)']))
    print
    print 'etas->V->Ds =',fit.p['Vnn'][0,0].fmt()
    print 'etas->V->Dso =',fit.p['Vno'][0,0].fmt()
    print
    ## error budget ##
    outputs = dict(Vnn=fit.p['Vnn'][0,0],Vno=fit.p['Vno'][0,0],
                 Eetas=dEetas[0],EDs=dEDs[0])
    inputs = {'stat.':[data[k] for k in data]} # statistical errors in data
    inputs.update(prior)                       # all entries in prior
    print fit.fmt_partialsdev(outputs,inputs,ndigit=3)
    ##
##

def fmtlist(x):
    return '  '.join([xi.fmt() for xi in x])
##
    
def build_prior(nexp):
    """ build prior """
    prior = BufferDict()
    ## etas ##
    prior.add('log(etas:a)',[log(gvar(0.3,0.3)) for i in range(nexp)])
    prior.add('log(etas:dE)',[log(gvar(0.5,0.5)) for i in range(nexp)])
    prior['log(etas:dE)'][0] = log(gvar(0.4,0.2))
    ##
    ## Ds ##
    prior.add('log(Ds:a)',[log(gvar(0.3,0.3)) for i in range(nexp)])
    prior.add('log(Ds:dE)',[log(gvar(0.5,0.5)) for i in range(nexp)])
    prior['log(Ds:dE)'][0] = log(gvar(1.2,0.2))
    prior.add('log(Ds:ao)',[log(gvar(0.1,0.1)) for i in range(nexp)])
    prior.add('log(Ds:dEo)',[log(gvar(0.5,0.5)) for i in range(nexp)])
    prior['log(Ds:dEo)'][0] = log(exp(prior['log(Ds:dE)'][0])+gvar(0.3,0.3))
    ##
    ## V ##
    prior.add('Vnn',[[gvar(0.1,1.0) for i in range(nexp)] for j in range(nexp)])
    prior.add('Vno',[[gvar(0.1,1.0) for i in range(nexp)] for j in range(nexp)])
    return prior
##

def build_models():
    """ build models """
    tmin = 5
    tdata = range(64)
    tfit = range(tmin,64-tmin) # all ts
    
    tp = 64 # periodic
    models = [
        Corr2(datatag='etas',tp=tp,tdata=tdata,tfit=tfit,
            a='log(etas:a)',b='log(etas:a)',dE='log(etas:dE)'),
            
        Corr2(datatag='Ds',tp=tp,tdata=tdata,tfit=tfit,
            a=('log(Ds:a)','log(Ds:ao)'),b=('log(Ds:a)','log(Ds:ao)'),
            dE=('log(Ds:dE)','log(Ds:dEo)'),s=(1.,-1.)),

        Corr3(datatag='3ptT15',tdata=range(16),T=15,tfit=range(tmin,16-tmin),
            a='log(etas:a)',dEa='log(etas:dE)',tpa=tp,
            b=('log(Ds:a)','log(Ds:ao)'),dEb=('log(Ds:dE)','log(Ds:dEo)'),
            tpb=tp,sb=(1,-1.),
            Vnn='Vnn',Vno='Vno',transpose_V=False),
            
        Corr3(datatag='3ptT16',tdata=range(17),T=16,tfit=range(tmin,17-tmin),
            a='log(etas:a)',dEa='log(etas:dE)',tpa=tp,
            b=('log(Ds:a)','log(Ds:ao)'),dEb=('log(Ds:dE)','log(Ds:dEo)'),
            tpb=tp,sb=(1,-1.),
            Vnn='Vnn',Vno='Vno',transpose_V=False)
    ]
    return models
##    

if __name__ == '__main__':
    main()

