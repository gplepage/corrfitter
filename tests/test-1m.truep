{"s:log(Ds:a)": [-1.537942027638434], "keys": ["s:log(etas:a)", "s:log(etas:dE)", "s:log(Ds:a)", "s:log(Ds:dE)", "s:log(Ds:ao)", "s:log(Ds:dEo)", "s:Vnn", "s:Vno"], "s:log(Ds:dEo)": [0.37734961251846294], "s:Vnn": [[0.769400211332711]], "s:log(etas:a)": [-1.5212974718929124], "s:log(Ds:dE)": [0.1837665525704063], "s:Vno": [[-0.7413349278338969]], "s:log(etas:dE)": [-0.876527586479904], "s:log(Ds:ao)": [-2.5965098875381902]}