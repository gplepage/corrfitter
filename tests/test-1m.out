========================== nexp = 1   ratio = True
Least Square Fit (input data correlated with prior):
  chi2/dof [dof] = 0.76 [69]    Q = 0.93    logGBF = -74.778    itns = 1

Parameters:
  log(etas:a) 0   -1.52130 (55)      [ -1.2 (1.0) ]
 log(etas:dE) 0   -0.87653 (20)      [ -0.92 (50) ]
    log(Ds:a) 0   -1.53794 (97)      [ -1.2 (1.0) ]
   log(Ds:dE) 0    0.18377 (12)      [  0.18 (17) ]
   log(Ds:ao) 0     -2.597 (23)      [ -2.3 (1.0) ]
  log(Ds:dEo) 0     0.3773 (32)      [  0.41 (24) ]
        Vnn 0,0     0.7694 (14)      [  0.1 (1.0) ]
        Vno 0,0     -0.741 (19)      [  0.1 (1.0) ]

Settings:
  svdcut = (None,None)   svdnum = (None,None)    reltol/abstol = 1e-10/1e-10

etas:dE = 0.416226(84)
 etas:E = 0.416226(84)  0.416226(84)  0.416226(84)
 etas:a = 0.21843(12)

Ds:dE = 1.20174(14)
 Ds:E = 1.20174(14)  1.20174(14)  1.20174(14)
 Ds:a = 0.21482(21)

etas->V->Ds = 0.7694(14)
etas->V->Dso = -0.741(19)

Partial % Errors:
                           Vno       EDs     Eetas       Vnn
------------------------------------------------------------
          log(Ds:a):     0.206     0.001     0.000     0.013
        log(etas:a):     0.022     0.000     0.001     0.015
         log(Ds:ao):     0.750     0.001     0.000     0.002
                Vno:     1.347     0.001     0.000     0.045
                Vnn:     0.435     0.000     0.001     0.072
        log(Ds:dEo):     0.448     0.001     0.000     0.008
       log(etas:dE):     0.045     0.000     0.002     0.018
         log(Ds:dE):     0.088     0.002     0.000     0.025
              stat.:     1.916     0.012     0.020     0.155
------------------------------------------------------------
              total:     2.547     0.012     0.020     0.181




========================== nexp = 1   ratio = False
Least Square Fit (input data correlated with prior):
  chi2/dof [dof] = 0.9 [69]    Q = 0.71    logGBF = -77.234    itns = 1

Parameters:
  log(etas:a) 0   -1.52144 (58)      [ -1.2 (1.0) ]
 log(etas:dE) 0   -0.87657 (20)      [ -0.92 (50) ]
    log(Ds:a) 0    -1.5379 (10)      [ -1.2 (1.0) ]
   log(Ds:dE) 0    0.18379 (12)      [  0.18 (17) ]
   log(Ds:ao) 0     -2.583 (28)      [ -2.3 (1.0) ]
  log(Ds:dEo) 0     0.3794 (35)      [  0.41 (24) ]
        Vnn 0,0    0.76569 (82)      [  0.1 (1.0) ]
        Vno 0,0     -0.725 (12)      [  0.1 (1.0) ]

Settings:
  svdcut = (None,None)   svdnum = (None,None)    reltol/abstol = 1e-10/1e-10

etas:dE = 0.416210(85)
 etas:E = 0.416210(85)  0.416210(85)  0.416210(85)
 etas:a = 0.21840(13)

Ds:dE = 1.20176(15)
 Ds:E = 1.20176(15)  1.20176(15)  1.20176(15)
 Ds:a = 0.21484(22)

etas->V->Ds = 0.76569(82)
etas->V->Dso = -0.725(12)

Partial % Errors:
                           Vno       EDs     Eetas       Vnn
------------------------------------------------------------
          log(Ds:a):     0.126     0.001     0.000     0.002
        log(etas:a):     0.026     0.000     0.001     0.003
         log(Ds:ao):     0.245     0.000     0.000     0.013
                Vno:     0.717     0.001     0.001     0.034
                Vnn:     0.255     0.001     0.000     0.028
        log(Ds:dEo):     0.110     0.000     0.000     0.014
       log(etas:dE):     0.022     0.000     0.002     0.002
         log(Ds:dE):     0.082     0.002     0.000     0.001
              stat.:     1.408     0.012     0.020     0.095
------------------------------------------------------------
              total:     1.630     0.012     0.020     0.107




